File: tree/binarytree/c/avl.c
#include "bstree.h"
#include <stdlib.h>

#define MAX(x, y) ((x) > (y) ? (x) : (y))
#define QLEN 1024

typedef struct Node {
  int val;
  size_t height;
  struct Node *left;
  struct Node *right;
} Node;

struct AVL {
  Node *root;
};

AVL* avl_create(void) {
  AVL *new = (AVL*) malloc(sizeof(AVL));
  if (new != NULL) {
    new->root = NULL;
  }
  return new;
}

static inline size_t height(Node *node) {
  if (node == NULL) {
    return 0;
  }

  size_t h1 = node->left == NULL ? 0 : node->left->height;
  size_t h2 = node->right == NULL ? 0 : node->right->height;

  return MAX(h1, h2) + 1;
}

static inline Node* LL(Node *node) {
  Node *right = node->right;

  node->right = right->left;
  right->left = node;

  node->height = height(node);
  return right;
}

static inline Node* RR(Node *node) {
  Node *left = node->left;

  node->left = left->right;
  left->right = node;

  node->height = height(node);
  return left;
}

static inline int balance_factor(Node *node) {
  return (int) height(node->left) - height(node->right);
}

static inline Node* balance(Node *node) {
  if (node == NULL) {
    return NULL;
  }

  if (balance_factor(node) > 1) {
    if (balance_factor(node->left) < 0) {
      node->left = LL(node->left);
    }
    node = RR(node);
  }
  else if (balance_factor(node) < -1) {
    if (balance_factor(node->right) > 0) {
      node->right = RR(node->right);
    }
    node = LL(node);
  }

  node->height = height(node);
  return node;
}

static inline Node* node_create(int val) {
  Node *new = (Node*) malloc(sizeof(Node));
  if (new != NULL) {
    new->val = val;
    new->height = 1;
    new->left = new->right = NULL;
  }
  return new;
}

static Node* insert(Node *node, int val) {
  if (node == NULL) {
    return node_create(val);
  }

  if (val < node->val) {
    node->left = insert(node->left, val);
  }
  else if (val > node->val) {
    node->right = insert(node->right, val);
  }

  return balance(node);
}

void avl_insert(AVL *avl, int val) {
  avl->root = insert(avl->root, val);
}

static Node* remove(Node *root, int val) {
  if (root == NULL) {
    return NULL;
  }

  if (val < root->val) {
    root->left = remove(root->left, val);
    return balance(root);
  }
  if (val > root->val) {
    root->right = remove(root->right, val);
    return balance(root);
  }

  if (root->left == NULL || root->right == NULL) {
    Node *tmp = root->left == NULL ? root->right : root->left;
    free(root);
    return tmp;
  }

  Node *n = root->left;
  while (n->right != NULL) {
    n = n->right;
  }

  root->val = n->val;
  root->left = remove(root->left, n->val);
  return balance(root);
}

void avl_remove(AVL *avl, int val) {
  avl->root = remove(avl->root, val);
}

bool avl_search(AVL *avl, int val) {
  Node *n = avl->root;
  while (n != NULL) {
    if (val == n->val) {
      return true;
    }
    n = val < n->val ? n->left : n->right;
  }
  return false;
}

static void node_free(Node *node) {
  if (node == NULL) {
    return;
  }
  node_free(node->left);
  node_free(node->right);
  free(node);
}

void avl_free(AVL *avl) {
  node_free(avl->root);
  free(avl);
}

static void preoder(Node *node, void (*consumer)(int val)) {
  if (node == NULL) {
    return;
  }
  consumer(node->val);
  preoder(node->left, consumer);
  preoder(node->right, consumer);
}

static void inorder(Node *node, void (*consumer)(int val)) {
  if (node == NULL) {
    return;
  }
  inorder(node->left, consumer);
  consumer(node->val);
  inorder(node->right, consumer);
}

static void postorder(Node *node, void (*consumer)(int val)) {
  if (node == NULL) {
    return;
  }
  postorder(node->left, consumer);
  postorder(node->right, consumer);
  consumer(node->val);
}

static void bfs(Node *root, void (*consumer)(int val)) {
  if (root == NULL) {
    return;
  }

  Node *queue[QLEN] = {root};
  size_t begin = 0;
  size_t len = 1;

  while (len > 0) {
    Node *node = queue[begin];
    len--;

    consumer(node->val);

    if (len < QLEN && node->left != NULL) {
      size_t index = (begin + len + 1) % QLEN;
      queue[index] = node->left;
      len++;
    }
    if (len < QLEN && node->right != NULL) {
      size_t index = (begin + len + 1) % QLEN;
      queue[index] = node->right;
      len++;
    }

    begin = (begin + 1) % QLEN;
  }
}

void avl_traversal(AVL *avl, Order order, void (*consumer)(int val)) {
  switch (order) {
    case ORDER_PRE:
      preoder(avl->root, consumer);
      return;
    case ORDER_IN:
      inorder(avl->root, consumer);
      return;
    case ORDER_POST:
      postorder(avl->root, consumer);
      return;
    case ORDER_BFS:
      bfs(avl->root, consumer);
  }
}

static size_t nodes(Node *node) {
  return node == NULL ? 0 : nodes(node->left) + nodes(node->right) + 1;
}

size_t avl_nodes(AVL *avl) {
  return nodes(avl->root);
}

static size_t levels(Node *node, size_t level) {
  return node == NULL ? 0 : MAX(levels(node->left, level + 1), levels(node->right, level + 1));
}

size_t avl_levels(AVL *avl) {
  return levels(avl->root, 0);
}

File: tree/binarytree/java/AVL.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.Stack;
import java.util.function.Consumer;

public class AVL implements BSTree {
    private Node root;

    public AVL() {
        this.root = null;
    }

    @Override
    public boolean equals(Object obj) {
        return this == obj || (obj instanceof AVL t && equals(root, t.root));
    }

    private boolean equals(Node t1, Node t2) {
        if (t1 == null || t2 == null) {
            return t1 == t2;
        }
        return t1.val == t2.val && equals(t1.left, t2.left) && equals(t1.right, t2.right);
    }

    @Override
    public String toString() {
        return buildList(root, Order.BFS).toString();
    }

    private List<Integer> buildList(Node root, Order order) {
        List<Integer> list = new ArrayList<>();

        if (root == null) {
            return list;
        }

        switch (order) {
            case PRE -> {
                Stack<Node> stack = new Stack<>();
                stack.push(root);
                while (!stack.isEmpty()) {
                    Node n = stack.pop();
                    list.add(n.val);

                    if (n.right != null) {
                        stack.push(n.right);
                    }
                    if (n.left != null) {
                        stack.push(n.left);
                    }
                }
            }
            case IN -> {
                Stack<Node> stack = new Stack<>();
                Node n = root;
                while (n != null || !stack.isEmpty()) {
                    while (n != null) {
                        stack.push(n);
                        n = n.left;
                    }
                    n = stack.pop();
                    list.add(n.val);
                    n = n.right;
                }
            }
            case POST -> {
                Stack<Node> stack = new Stack<>();
                stack.push(root);
                while (!stack.isEmpty()) {
                    Node n = stack.pop();
                    list.addFirst(n.val);

                    if (n.left != null) {
                        stack.push(n.left);
                    }
                    if (n.right != null) {
                        stack.push(n.right);
                    }
                }
            }
            case BFS -> {
                Deque<Node> deque = new ArrayDeque<>();
                deque.add(root);
                while (!deque.isEmpty()) {
                    Node n = deque.removeFirst();
                    list.add(n.val);
                    if (n.left != null) {
                        deque.add(n.left);
                    }
                    if (n.right != null) {
                        deque.add(n.right);
                    }
                }
            }
        }

        return list;
    }

    @Override
    public void traversal(Order order, Consumer<Integer> consumer) {
        buildList(root, order).forEach(consumer);
    }

    private int height(Node node) {
        if (node == null) {
            return 0;
        }

        int h1 = node.left == null ? 0 : node.left.height;
        int h2 = node.right == null ? 0 : node.right.height;

        return Math.max(h1, h2) + 1;
    }

    private Node LL(Node node) {
        Node right = node.right;

        node.right = right.left;
        right.left = node;

        node.height = height(node);
        return right;
    }

    private Node RR(Node node) {
        Node left = node.left;

        node.left = left.right;
        left.right = node;

        node.height = height(node);
        return left;
    }

    private int balance_factor(Node node) {
        return node == null ? 0 : height(node.left) - height(node.right);
    }

    private Node balance(Node node) {
        if (node == null) {
            return null;
        }

        if (balance_factor(node) > 1) {
            if (balance_factor(node.left) < 0) {
                node.left = LL(node.left);
            }
            node = RR(node);
        }
        if (balance_factor(node) < -1) {
            if (balance_factor(node.right) > 0) {
                node.right = RR(node.right);
            }
            node = LL(node);
        }

        node.height = height(node);
        return node;
    }

    @Override
    public void insert(int val) {
        root = insert(root, val);
    }

    private Node insert(Node root, int val) {
        if (root == null) {
            return new Node(val);
        }

        if (val < root.val) {
            root.left = insert(root.left, val);
        }
        if (val > root.val) {
            root.right = insert(root.right, val);
        }

        return balance(root);
    }

    @Override
    public int levels() {
        return levels(root, 0);
    }

    private int levels(Node root, int total) {
        return root == null ? total : Math.max(levels(root.left, total + 1), levels(root.right, total + 1));
    }

    @Override
    public int nodes() {
        return nodes(root, 0);
    }

    private int nodes(Node root, int total) {
        return root == null ? 0 : nodes(root.left, total) + nodes(root.right, total) + 1;
    }

    @Override
    public void remove(int val) {
        root = remove(root, val);
    }

    private Node remove(Node root, int val) {
        if (root == null) {
            throw new NoSuchElementException(String.valueOf(val));
        }

        if (val < root.val) {
            root.left = remove(root.left, val);
            return balance(root);
        }
        if (val > root.val) {
            root.right = remove(root.right, val);
            return balance(root);
        }

        if (root.left == null) {
            root = root.right;
            return root;
        }
        if (root.right == null) {
            root = root.left;
            return root;
        }

        Node n = root.left;
        while (n.right != null) {
            n = n.right;
        }

        root.val = n.val;
        root.left = remove(root.left, n.val);
        return balance(root);
    }

    @Override
    public boolean search(int val) {
        Node n = root;

        while (n != null) {
            if (val == n.val) {
                return true;
            }
            n = val < n.val ? n.left : n.right;
        }

        return false;
    }

    private static class Node {
        int val;
        Node left;
        Node right;
        int height;

        Node(int val, Node left, Node right) {
            this.val = val;
            this.left = left;
            this.right = right;
            this.height = 1;
        }

        Node(int val) {
            this(val, null, null);
        }
    }
}
